### PR TITLE
Added ability to replace ARC machines, fixes disconnect and reports errors

### PR DESCRIPTION
**Pull request by strottos**
If the `ForceReplaceAgent` parameter is specified and the agent isn't
connected it will check for any existing agents in the same resource
group and subscription with the same name and if it finds any it will
delete them from Azure using the Arc credentials specified and the Azure
REST API.

Fixes an error in the disconnect step where it was previously supplying
the wrong parameters to `azcmagent`.

Reports any errors to stdout now so DSC reports them to the end user.
Previously they were being swallowed by the process itself and it wasn't
clear why things were failing.

### CODE REVIEW

**Comment by strottos** [comment]
Fixes issue #2

Has been tested on the system it was written for and seems to be working well there. Also all unit and integration tests are passing (with new ones added in the PR).

Anything anyone needs with this please let me know.

**Comment by richeney** [comment]
@mgreenegit

Hi Michael - any chance this could be reviewed at pace? It is contributed by one of the UK partners as it is a blocker for work they are doing with their customer.

**Comment by richeney** [comment]
@mgreenegit

Hi Michael - any chance this could be reviewed at pace? It is contributed by one of the UK partners as it is a blocker for work they are doing with Rare.

**Comment by strottos** [comment]
Any word on when this might get looked at? And anything I can do to help push this along?

**Comment by mgreenegit** [comment]
I like this change. I will set aside time tomorrow to do a review and test. Sorry I didn't see this earlier. Thank you!
Also, after we merge this change, I will try and update it and hand it off to the community so our reaction time will be better.

**Comment by strottos** [comment]
Without meaning to be pushy, anything further on this? Appreciate any time you can find for this, I realise this likely isn't priority. Happy to help in any way I can.

**Comment by mgreenegit** [comment]
@strottos sorry it has taken so long to get back to you (> 9 months).

I reviewed this in depth with @rpsqrd. Our only concern is that the SPN used to register Arc machines should follow least privilege, and only have permission to register new machines. If the ID also has permission to delete machines and someone harvests it from the automation, it could be used inappropriately (be it accidentally or maliciously).

Another option would be to specify a unique name for the machine using the -n parameter for azcmagent. Some external automation would need to be used then (Functions, possibly) to "tombstone" and then delete old resources. This is also a scenario we should consider for the service backlog, in my humble opinion.

The other parts of this PR are really valuable. It is always best to capture errors and output, exactly as you have provided. The disconnect fix is also a great catch.

I am going to be refactoring this resource to be class-based and adding a new resource to the module for features of the latest agent.

Would you prefer to update the PR, or close it and I can comment credit to your in a new release with the disconnect fix, and the error output?

cc @richeney

**Comment by strottos** [comment]
I did get quite into this as you can probably see, but it's a while ago so I'm happy to hand over control to you to fix these things. I can certainly remember it didn't log errors properly so I made it do so in order to figure out my disconnect issues. I would certainly recommend taking these fixes as the problem below would just show as a "an error occurred" kind of error message but with these fixes it gave fairly detailed errors.

The disconnect problem was the main issue for us but we've managed by manually deleting basically. The use case was for on premise servers that occasionally (well frequently actually given the number of servers but anyway) stop behaving and the fix has always been to reimage them. Then when the DSC module here kicks in it tries to register an Arc machine with the same name, unfortunately as one already exists Azure says no. Having thought about it after this PR I became convinced this is probably more an issue with Arc and/or this process than this module so happy for that not to be done. It's a further reaching problem than just getting it to register with Arc IIRC when it's a duplicate machine name.

tl;dr Please go ahead and take the fixes you like as you see best, all good with me 👍

**Comment by mgreenegit** [comment]
Got it. I will close this PR and I have commented your alias in the commit history and change log to give you credit.
